lib/utils/hostname.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"golang.org/x/exp/slices"
+)
+
+func isFQDN(hostname string) bool {
+	return strings.Contains(hostname, ".")
+}
+
+func isWildcard(hostname string) bool {
+	return strings.HasPrefix(hostname, "*.")
+}
+
+func wildcardFromHostname(hostname string) string {
+	if isWildcard(hostname) {
+		return hostname
+	}
+	return fmt.Sprintf("*.%s", strings.Join(strings.Split(hostname, ".")[1:], "."))
+}
+
+// AddHostToHostList adds the provided hostname to the hostList, then processes the hostList to add wildcards
+// if appropriate. it returns a processed, sorted hostList.
+//
+// for every host in the list:
+//   - all hostnames are converted to their wildcard form (test.example.com -> *.example.com) and grouped together.
+//   - if we only have a single hostname which would be matched by a given wildcard, that hostname is added
+//     to the hostList verbatim to prevent matching against too many hosts with a wildcard.
+//   - if a wildcard matches multiple hostnames, the wildcard will be added to the hostList instead.
+//
+// this is an effort to keep the length of hostList as short as possible for efficiency and tidiness.
+func AddHostToHostList(hostList []string, hostname string) []string {
```

Comment:
`utils` packages are generally frowned upon. This seems like pretty specific logic that would make sense in a dedicated package with proper types.

Just looking at the godoc for this, I would have no idea when it's appropriate to use, and the fact that it's in a `utils` package doesn't give me any other clues. When is it appropriate to add wildcards?


I would consider:
- putting this in a package with a more descriptive name (for example, if this format is specific to putty, maybe `package puttyhosts`)
- making a proper data type. perhaps `type HostList struct`, with a method to add a host `func (h *HostList) AddHost` and a method to get the completed list.

Reply:
I've moved the functions into a `puttyhosts` package as suggested and improved the Godoc considerably to add background on why the function is needed, what PuTTY uses it its output for, its expected usage and exactly what it does. The unit test adds good context on its job as well.

I'm happy to go further with refining this into a `type` instead if appropriate, but I'm hopeful that for the relatively limited usage this function will probably see, this will suffice.